dataset/bin.cpp

```diff
@@ -93,9 +96,11 @@ template <class T> Variable as_subspan_view(T &&binned) {
 /// indices is a binned variable with sub-bin indices, i.e., new bins within
 /// bins
 Variable bin_sizes(const VariableConstView &sub_bin, const scipp::index nbin) {
-  const auto nbins = broadcast(nbin * units::one, sub_bin.dims());
-  auto sizes = resize(sub_bin, nbins);
-  buckets::reserve(sizes, nbins);
+  const auto end = cumsum(broadcast(nbin * units::one, sub_bin.dims()));
+  const auto dim = variable::variableFactory().elem_dim(sub_bin);
+  auto sizes = make_bins(
+      zip(end - nbin * units::one, end), dim,
+      makeVariable<scipp::index>(Dims{dim}, Shape{end.dims().volume() * nbin}));
```

Comment:
I can't tell, is this code doing something different than before? Could this affect the performance?

Reply:
It is doing the same, but in a different manner:
- Previously it took `sub_bin` which was a binned variable and resized the bins to `nbin`.
- Now `sub_bin` may have (inner) dtype `int32_t` but the event *count* is always stored as `int64_t`/`scipp::index`. I could have resized and used `astype`, but the new way felt more natural.

It is not related to the speedup, since I originally overlooked that I had to change the type here and saw the same results.

The performance-relevant part here is that `sub_bin` is now typically `int32_t`. Its size is the total number of events, so this matters (it is a part of the overall speedup, but definitely not the only). It accumulates into `int_64` (one for each bin).